From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Gregory Morse <gregory.morse@live.com>
Date: Sun, 28 Jul 2019 17:35:17 +0200
Subject: [PATCH] 0852: Double segmented resolve bad output in constants

Instead of printing the global variable, relics like "ram0x000006d0" occur because the resolver is run twice, the first time on a near/far pointer, the second time on a RAM address which it treats like a segmented far pointer due to its size - and this is wrong behavior.  Yet the ptr check flag is set specifically to prevent this yet never inquired.

Update type.cc
Update printc.cc
Update coreaction.cc
Update ruleaction.cc
---
 .../Decompiler/src/decompile/cpp/coreaction.cc         | 10 +++++-----
 .../Decompiler/src/decompile/cpp/funcdata_varnode.cc   |  2 +-
 Ghidra/Features/Decompiler/src/decompile/cpp/printc.cc |  2 +-
 .../Decompiler/src/decompile/cpp/ruleaction.cc         |  2 +-
 Ghidra/Features/Decompiler/src/decompile/cpp/type.cc   |  3 ++-
 Ghidra/Features/Decompiler/src/decompile/cpp/type.hh   |  2 +-
 6 files changed, 11 insertions(+), 10 deletions(-)

diff --git a/Ghidra/Features/Decompiler/src/decompile/cpp/coreaction.cc b/Ghidra/Features/Decompiler/src/decompile/cpp/coreaction.cc
index f59f259998..2ce3cddc96 100644
--- a/Ghidra/Features/Decompiler/src/decompile/cpp/coreaction.cc
+++ b/Ghidra/Features/Decompiler/src/decompile/cpp/coreaction.cc
@@ -4983,13 +4983,13 @@ void ActionInputPrototype::gatherParamSpacebaseRefs(list<InputRef> &refs,Funcdat
     OpCode opc = op->code();
     if (opc == CPUI_PTRSUB) {
       vn = op->getIn(1);
-      addr = sbtype->getAddress(vn->getOffset(),vn->getSize(),op->getAddr());
+      addr = sbtype->getAddress(vn->getOffset(),vn->getSize(),op->getAddr(),vn->isPtrCheck());
     }
     else if (opc == CPUI_PTRADD) {
       vn = op->getIn(1);
       if (vn->isConstant()) {
 	uintb off = vn->getOffset() * op->getIn(2)->getOffset();
-	addr = sbtype->getAddress(off,vn->getSize(),op->getAddr());
+	addr = sbtype->getAddress(off,vn->getSize(),op->getAddr(),vn->isPtrCheck());
       }
       else
 	continue;
@@ -5653,14 +5653,14 @@ void ActionInferTypes::propagateSpacebaseRef(Funcdata &data,Varnode *spcvn)
     switch(op->code()) {
     case CPUI_COPY:
       vn = op->getIn(0);
-      addr = sbtype->getAddress(0,vn->getSize(),op->getAddr());
+      addr = sbtype->getAddress(0,vn->getSize(),op->getAddr(),vn->isPtrCheck());
       propagateRef(data,op->getOut(),addr);
       break;
     case CPUI_INT_ADD:
     case CPUI_PTRSUB:
       vn = op->getIn(1);
       if (vn->isConstant()) {
-	addr = sbtype->getAddress(vn->getOffset(),vn->getSize(),op->getAddr());
+	addr = sbtype->getAddress(vn->getOffset(),vn->getSize(),op->getAddr(),vn->isPtrCheck());
 	propagateRef(data,op->getOut(),addr);
       }
       break;
@@ -5668,7 +5668,7 @@ void ActionInferTypes::propagateSpacebaseRef(Funcdata &data,Varnode *spcvn)
       vn = op->getIn(1);
       if (vn->isConstant()) {
 	uintb off = vn->getOffset() * op->getIn(2)->getOffset();
-	addr = sbtype->getAddress(off,vn->getSize(),op->getAddr());
+	addr = sbtype->getAddress(off,vn->getSize(),op->getAddr(),vn->isPtrCheck());
 	propagateRef(data,op->getOut(),addr);
       }
       break;
diff --git a/Ghidra/Features/Decompiler/src/decompile/cpp/funcdata_varnode.cc b/Ghidra/Features/Decompiler/src/decompile/cpp/funcdata_varnode.cc
index 9b650aa50c..155680e9aa 100644
--- a/Ghidra/Features/Decompiler/src/decompile/cpp/funcdata_varnode.cc
+++ b/Ghidra/Features/Decompiler/src/decompile/cpp/funcdata_varnode.cc
@@ -1272,7 +1272,7 @@ Symbol *Funcdata::linkSymbolReference(Varnode *vn)
   if (sb->getMetatype() != TYPE_SPACEBASE)
       return (Symbol *)0;
   Scope *scope = sb->getMap();
-  Address addr = sb->getAddress(vn->getOffset(),in0->getSize(),op->getAddr());
+  Address addr = sb->getAddress(vn->getOffset(),in0->getSize(),op->getAddr(),vn->isPtrCheck());
   if (addr.isInvalid())
     throw LowlevelError("Unable to generate proper address from spacebase");
   MapEntry *entry = scope->queryContainer(addr,1,Address());
diff --git a/Ghidra/Features/Decompiler/src/decompile/cpp/printc.cc b/Ghidra/Features/Decompiler/src/decompile/cpp/printc.cc
index 6ccbfd1a98..dbf10d2180 100644
--- a/Ghidra/Features/Decompiler/src/decompile/cpp/printc.cc
+++ b/Ghidra/Features/Decompiler/src/decompile/cpp/printc.cc
@@ -1096,7 +1096,7 @@ void PrintC::opPtrsub(const PcodeOp *op)
     }
     if (symbol == (Symbol *)0) {
       TypeSpacebase *sb = (TypeSpacebase *)ct;
-      Address addr = sb->getAddress(in1const,in0->getSize(),op->getAddr());
+      Address addr = sb->getAddress(in1const,in0->getSize(),op->getAddr(),op->getIn(1)->isPtrCheck());
       pushUnnamedLocation(addr,(Varnode *)0,op);
     }
     else {
diff --git a/Ghidra/Features/Decompiler/src/decompile/cpp/ruleaction.cc b/Ghidra/Features/Decompiler/src/decompile/cpp/ruleaction.cc
index 1a638ebcc7..80c0d8f099 100644
--- a/Ghidra/Features/Decompiler/src/decompile/cpp/ruleaction.cc
+++ b/Ghidra/Features/Decompiler/src/decompile/cpp/ruleaction.cc
@@ -7312,7 +7312,7 @@ int4 RulePtrsubCharConstant::applyOp(PcodeOp *op,Funcdata &data)
   if (outtype->getMetatype() != TYPE_PTR) return 0;
   Datatype *basetype = outtype->getPtrTo();
   if (!basetype->isCharPrint()) return 0;
-  Address symaddr = sbtype->getAddress(vn1->getOffset(),vn1->getSize(),op->getAddr());
+  Address symaddr = sbtype->getAddress(vn1->getOffset(),vn1->getSize(),op->getAddr(),vn1->isPtrCheck());
   Scope *scope = sbtype->getMap();
   if (!scope->isReadOnly(symaddr,1,op->getAddr()))
     return 0;
diff --git a/Ghidra/Features/Decompiler/src/decompile/cpp/type.cc b/Ghidra/Features/Decompiler/src/decompile/cpp/type.cc
index 6b226036cb..ff65aae26e 100644
--- a/Ghidra/Features/Decompiler/src/decompile/cpp/type.cc
+++ b/Ghidra/Features/Decompiler/src/decompile/cpp/type.cc
@@ -3674,10 +3674,11 @@ bool TypeSpacebase::isOffsetValid(int8 off,int8 extra,int8 multiplier) const
 /// \param sz is the size of offset (as a pointer)
 /// \param point is a "context" reference for the request
 /// \return the referred to Address
-Address TypeSpacebase::getAddress(uintb off,int4 sz,const Address &point) const
+Address TypeSpacebase::getAddress(uintb off,int4 sz,const Address &point,bool ptrCheck) const
 
 {
   uintb fullEncoding;
+  if (ptrCheck) return Address(spaceid,off);
   // Currently a constant off of a global spacebase must be a full pointer encoding
   if (scopeId == 0)
     sz = -1;	// Set size to -1 to guarantee that full encoding recovery isn't launched
diff --git a/Ghidra/Features/Decompiler/src/decompile/cpp/type.hh b/Ghidra/Features/Decompiler/src/decompile/cpp/type.hh
index b32741b41c..c094796e54 100644
--- a/Ghidra/Features/Decompiler/src/decompile/cpp/type.hh
+++ b/Ghidra/Features/Decompiler/src/decompile/cpp/type.hh
@@ -832,7 +832,7 @@ public:
   TypeSpacebase(AddrSpace *id,uint8 scope,Architecture *g)
     : Datatype(0,1,TYPE_SPACEBASE) { spaceid = id; glb = g; scopeId = scope; }
   Scope *getMap(void) const;	///< Get the symbol table indexed by \b this
-  Address getAddress(uintb off,int4 sz,const Address &point) const;	///< Construct an Address given an offset
+  Address getAddress(uintb off,int4 sz,const Address &point,bool ptrCheck) const;	///< Construct an Address given an offset
   virtual Datatype *getSubType(int8 off,int8 *newoff) const;
   virtual bool nearestArrayedComponentForward(int8 off,Nearest &res) const;
   virtual bool nearestArrayedComponentBackward(int8 off,Nearest &res) const;
-- 
2.45.1

